include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -560,24 +584,45 @@ class permutation_iterator
 
     permutation_iterator() = default;
 
-    permutation_iterator(const SourceIterator& input1, const IndexMap& input2, ::std::size_t index = 0)
-        : my_source_it(input1), my_index_map(input2), my_index(index)
+    template <typename _T = _Permutation, typename ::std::enable_if<!__internal::__is_functor<_T>, int>::type = 0>
+    permutation_iterator(SourceIterator input1, _Permutation input2) : my_source_it(input1), my_index(input2)
     {
     }
 
+    template <typename _T = _Permutation, typename ::std::enable_if<__internal::__is_functor<_T>, int>::type = 0>
+    permutation_iterator(SourceIterator input1, _Permutation __f, difference_type __idx = 0)
+        : my_source_it(input1), my_index(counting_iterator<difference_type>(__idx), __f)
+    {
+    }
+
+  private:
+    template <typename _T = _Permutation, typename ::std::enable_if<__internal::__is_functor<_T>, int>::type = 0>
+    permutation_iterator(SourceIterator input1, IndexMap input2) : my_source_it(input1), my_index(input2)
+    {
+    }
+
+  public:
     SourceIterator
     base() const
     {
         return my_source_it;
     }
 
+    template <typename _T = IndexMap, typename ::std::enable_if<!__internal::__is_functor<_T>, int>::type = 0>
+    IndexMap
+    map() const
+    {
+        return my_index;
+    }
+
+    template <typename _T = IndexMap, typename ::std::enable_if<__internal::__is_functor<_T>, int>::type = 0>
     IndexMap
     map() const
     {
-        return my_index_map + my_index;
+        return my_index.functor();
     }
```

Comment:
`IndexMap` seems to be a transform iterator (line 572-573), but the return value is a functor. Looks like an error.

Reply:
Actually, we have two "SFINAE specializations"  of `map()`...
But... we keep just one and use `std::conditional` in the "return" line. It will be more readable and shorter.